plugin/path_login.go

```diff
@@ -298,6 +352,187 @@ func (b *GcpAuthBackend) validateAgainstIAMRole(serviceAccount *iam.ServiceAccou
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Period: role.Period,
+			Persona: &logical.Persona{
+				Name: fmt.Sprintf("gce-%s", strconv.FormatUint(instance.Id, 10)),
+			},
+			Policies: role.Policies,
+			Metadata: map[string]string{
+				"project_id":                  metadata.ProjectId,
+				"project_number":              strconv.FormatInt(metadata.ProjectNumber, 10),
+				"zone":                        metadata.Zone,
+				"instance_id":                 metadata.InstanceId,
+				"instance_name":               metadata.InstanceName,
+				"instance_creation_timestamp": metadata.CreatedAt,
+				"service_account_id":          loginInfo.ServiceAccountId,
+				"role":                        loginInfo.RoleName,
+			},
+			DisplayName: instance.Name,
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
+				TTL:       role.TTL,
+			},
+		},
+	}
+
+	return resp, nil
+}
+
+// pathGceRenew returns an error if the instance referenced in the auth token metadata cannot renew the
+// auth token for the given role.
+func (b *GcpAuthBackend) pathGceRenew(req *logical.Request, role *gcpRole) error {
+	gceClient, err := b.GCE(req.Storage)
+	if err != nil {
+		return fmt.Errorf(clientErrorTemplate, "GCE", err)
+	}
+
+	meta := &util.GCEIdentityMetadata{}
+	if err := mapstructure.WeakDecode(req.Auth.Metadata, meta); err != nil {
+		return fmt.Errorf("invalid auth metadata: %v", err)
+	}
+
+	instance, err := meta.GetVerifiedInstance(gceClient)
+	if err != nil {
+		return err
+	}
+
+	serviceAccountId, ok := req.Auth.Metadata["service_account_id"]
+	if !ok {
+		return errors.New("invalid auth metadata: service_account_id not found")
+	}
+	if err := b.authorizeGCEInstance(instance, req.Storage, role, meta.Zone, serviceAccountId); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateGCEInstance returns an error if the given GCE instance is not authorized for the role.
+func (b *GcpAuthBackend) authorizeGCEInstance(instance *compute.Instance, s logical.Storage, role *gcpRole, zone, serviceAccountId string) error {
+	gceClient, err := b.GCE(s)
+	if err != nil {
+		return err
+	}
+
+	// Verify instance has role labels if labels were set on role.
+	for k, expectedV := range role.Labels {
+		actualV, ok := instance.Labels[k]
+		if !ok || actualV != expectedV {
+			return fmt.Errorf("role label '%s:%s' not found on GCE instance", k, expectedV)
+		}
+	}
+
+	// Verify that instance is in zone or region if given.
+	if len(role.Zone) > 0 {
+		var zone string
+		idx := strings.LastIndex(instance.Zone, "zones/")
+		if idx > 0 {
+			// Parse zone name from full zone self-link URL.
+			idx += len("zones/")
+			zone = instance.Zone[idx:len(instance.Zone)]
+		} else {
+			// Expect full zone name to be set as instance zone.
+			zone = instance.Zone
+		}
+
+		if zone != role.Zone {
+			return fmt.Errorf("instance is not in role zone '%s'", role.Zone)
+		}
+	} else if len(role.Region) > 0 {
```

Comment:
Is this the proper behavior if both role.Zone and role.Region are set, or should these be independent if clauses?

Reply:
If Zone is set, we will ignore Region and use the region of the given Zone.